applications/plugins/SofaCUDA/sofa/gpu/cuda/CudaHexahedronFEMForceField.inl

```diff
@@ -0,0 +1,370 @@
+
+  }
+
+  m->_elementStiffnesses.endEdit();
+
+}
+
+template<class TCoord, class TDeriv, class TReal>
+void HexahedronFEMForceFieldInternalData< gpu::cuda::CudaVectorTypes<TCoord,TDeriv,TReal> >::addForce(Main* m, VecDeriv& f, const VecCoord& x, const VecDeriv& v)
+{
+  Data& data = *m->data;
+
+    f.resize(x.size());
+    Kernels::addForce(
+        data.GATHER_PT,
+        data.GATHER_BSIZE,
+        data.size(),
+        data.nbVertex,
+        data.nbElementPerVertex,
+        data.elems.deviceRead(),
+        data.erotation.deviceWrite(),
+        data.ekmatrix.deviceRead(),
+        data.eforce.deviceWrite(),
+        data.velems.deviceRead(),
+        (      Deriv*)f.deviceWrite() + data.vertex0,
+        (const Coord*)x.deviceRead()  + data.vertex0,
+        (const Deriv*)v.deviceRead()  + data.vertex0);
+} // addForce
+
+template<class TCoord, class TDeriv, class TReal>
+void HexahedronFEMForceFieldInternalData< gpu::cuda::CudaVectorTypes<TCoord,TDeriv,TReal> >::addDForce (Main* m, VecDeriv& df, const VecDeriv& dx, double kFactor/*, double bFactor*/)
+{
+    Data& data = *m->data;
+    df.resize(dx.size());
+
+    Kernels::addDForce(        
+        data.GATHER_PT,
+        data.GATHER_BSIZE,
+        data.size(),
+        data.nbVertex,
+        data.nbElementPerVertex,
+        data.elems.deviceRead(),
+        data.erotation.deviceRead(),
+        data.ekmatrix.deviceRead(),
+        data.eforce.deviceWrite(),
+        data.velems.deviceRead(),
+        (      Deriv*)df.deviceWrite() + data.vertex0,
+        (const Deriv*)dx.deviceRead()  + data.vertex0,
+         kFactor);
+
+} // addDForce
+
+
+//template<class TCoord, class TDeriv, class TReal>
+//void HexahedronFEMForceFieldInternalData< gpu::cuda::CudaVectorTypes<TCoord,TDeriv,TReal> >::getNodeRotations(Main* m, sofa::gpu::cuda::CudaRotationMatrix<TReal> * diag, int offset)
+//{
+//    Data& data = *m->data;
+//
+//    Kernels::getRotations(
+//        data.GATHER_PT,
+//        data.GATHER_BSIZE,
+//        data.nbVertex,
+//        data.nbElementPerVertex,
+//        data.velems.deviceRead(),
+//        data.erotation.deviceRead(),
+//        data.irotation.deviceRead(),
+//        diag->getVector().deviceWrite(offset*9));
+//
+//} // addDForce
+
+template<>
+void HexahedronFEMForceField< gpu::cuda::CudaVec3fTypes >::reinit()	{ data->reinit(this);}
+template<>
+void HexahedronFEMForceField< gpu::cuda::CudaVec3fTypes >::addForce(const core::MechanicalParams* /*mparams*/ /* PARAMS FIRST */, DataVecDeriv& d_f, const DataVecCoord& d_x, const DataVecDeriv& d_v)
+{ 
+	VecDeriv& f = *d_f.beginEdit();
+	const VecCoord& x = d_x.getValue();
+	const VecDeriv& v = d_v.getValue();
+
+	data->addForce(this, f, x, v); 
+
+	d_f.endEdit();
+}
+template<>
+void HexahedronFEMForceField< gpu::cuda::CudaVec3fTypes >::addDForce(const core::MechanicalParams* mparams /* PARAMS FIRST */, DataVecDeriv& d_df, const DataVecDeriv& d_dx)
+{ 
+	VecDeriv& df = *d_df.beginEdit();
```

Comment:
For SofaCUDA, VecDeriv is a ```CudaVector``` inheriting from ```vector_device``` with ```deviceRead()/deviceWrite()```  methods, which (if I'm not wrong) are not in the accessors. 
I tried to use them but it doesn't work. 
This could be reported in a issue for later.